### PR TITLE
[FIX] std::hash + view::kmer_hash with const 

### PR DESCRIPTION
**Pull request by eseiler**

### CODE REVIEW

File: include/seqan3/alphabet/detail/hash.hpp [hunk 1]
```diff
@@ -21,7 +21,8 @@ namespace std
  * \ingroup alphabet
  * \tparam alphabet_t The type of character to hash; Must model seqan3::Semialphabet.
  */
-template <seqan3::Semialphabet alphabet_t>
+template <typename alphabet_t>
+    requires seqan3::Semialphabet<seqan3::remove_cvref_t<alphabet_t>>
```
Comment (h-2) [inline include/seqan3/alphabet/detail/hash.hpp]: only remove const

File: include/seqan3/alphabet/detail/hash.hpp [hunk 2]
```diff
@@ -45,7 +46,7 @@ struct hash<alphabet_t>
  */
 template <ranges::InputRange urng_t>
     //!\cond
-    requires seqan3::Semialphabet<seqan3::reference_t<urng_t>>
+    requires seqan3::Semialphabet<seqan3::reference_t<seqan3::remove_cvref_t<urng_t>>>
```
Comment (h-2) [inline include/seqan3/alphabet/detail/hash.hpp]: only remove const and remove around the reference_t, not inside. There are non-const views that still have const-reference type, like std::basic_string_view.

File: include/seqan3/range/view/kmer_hash.hpp [hunk 3]
```diff
@@ -43,14 +43,14 @@ class kmer_hash_fn : public pipable_adaptor_base<kmer_hash_fn>
      */
     template <std::ranges::ViewableRange urng_t>
     //!\cond
-        requires Semialphabet<reference_t<urng_t>>
+        requires Semialphabet<reference_t<remove_cvref_t<urng_t>>>
```
Comment (h-2) [inline include/seqan3/range/view/kmer_hash.hpp]: same here

File: include/seqan3/range/view/kmer_hash.hpp [hunk 4]
```diff
@@ -43,14 +43,14 @@ class kmer_hash_fn : public pipable_adaptor_base<kmer_hash_fn>
...
     //!\endcond
     static auto impl(urng_t && urange, size_t const k)
     {
         return std::forward<urng_t>(urange) | ranges::view::sliding(k) | std::view::transform(
         [] (auto const in)
         {
-            std::hash<decltype(in)> h{};
+            std::hash<remove_cvref_t<urng_t>> h{};
```
Comment (h-2) [inline include/seqan3/range/view/kmer_hash.hpp]: why this change? `in` is an alphabet but you are hashing with the specialisation of a range 🤔

File: include/seqan3/alphabet/detail/hash.hpp [hunk 5]
```diff
@@ -21,7 +21,8 @@ namespace std
  * \ingroup alphabet
  * \tparam alphabet_t The type of character to hash; Must model seqan3::Semialphabet.
  */
-template <seqan3::Semialphabet alphabet_t>
+template <typename alphabet_t>
+    requires seqan3::Semialphabet<seqan3::delete_const_t<alphabet_t>>
```
Comment (smehringer) [inline include/seqan3/alphabet/detail/hash.hpp]: `//!\cond` missing

File: include/seqan3/core/metafunction/basic.hpp [hunk 6]
```diff
@@ -23,6 +23,59 @@ namespace seqan3
...
+ * \tparam t The type to operate on.
+ * \attention In contrast to std::remove_const, this function also removes `const` from pointers and references.
+ */
+template <typename t>
+struct delete_const
+{
+    //!\brief The return type is the input type with the topmost `const` stripped.
+    using type = t;
+};
+//
```
Comment (h-2) [inline include/seqan3/core/metafunction/basic.hpp]: ?

File: include/seqan3/core/metafunction/basic.hpp [hunk 7]
```diff
@@ -23,6 +23,59 @@ namespace seqan3
...
+//!\overload
```
Comment (h-2) [inline include/seqan3/core/metafunction/basic.hpp]: I don't think //!overload works for metafunctions... have you verified how the documentation looks? For simple metafunctions like this, if you can't make it appear nice, just hide the "overloads" with \cond and clearly document the behaviour in the main implementation

File: include/seqan3/core/metafunction/basic.hpp [hunk 8]
```diff
@@ -23,6 +23,59 @@ namespace seqan3
...
+template <typename t>
+struct delete_const<const t>
```
Comment (h-2) [inline include/seqan3/core/metafunction/basic.hpp]: we use east-const in SeqAn → '<t const>`

Comment (h-2) [reply include/seqan3/core/metafunction/basic.hpp]: also beloew

File: test/unit/core/metafunction/basic_test.cpp [hunk 9]
```diff
@@ -36,3 +36,19 @@ TEST(metafunction, remove_cvref_)
...
+    EXPECT_TRUE((std::is_same_v<int volatile, delete_const_t<int volatile>>));
+    EXPECT_TRUE((std::is_same_v<int &, delete_const_t<int &>>));
+    EXPECT_TRUE((std::is_same_v<int &&, delete_const_t<int &&>>));
+    EXPECT_TRUE((std::is_same_v<int &, delete_const_t<int const &>>));
+    EXPECT_TRUE((std::is_same_v<int &&, delete_const_t<int const &&>>));
+    EXPECT_TRUE((std::is_same_v<int volatile &&, delete_const_t<int const volatile &&>>));
+    EXPECT_TRUE((std::is_same_v<int *, delete_const_t<int *>>));
+    EXPECT_TRUE((std::is_same_v<int *, delete_const_t<int const *>>));
+    EXPECT_TRUE((std::is_same_v<int volatile *, delete_const_t<int const volatile *>>));
+    EXPECT_TRUE((std::is_same_v<char const *, delete_const_t<char const * const>>));
```
Comment (h-2) [inline test/unit/core/metafunction/basic_test.cpp]: Might be more readable if the two arguments are aligned vertically, but maybe that's just my taste
